File: androidplot-master/demoapp/src/main/java/methods/IncrementalSearch.java
package methods; /**
 * IncrementalSearch
 */
import methods.com.udojava.evalex.Expression;
import java.math.BigDecimal;
import java.util.ArrayList;


public class IncrementalSearch {
   private String function;
   private BigDecimal x0;
   private double delta;
   private double[] result;
   private int niter;
   private String message;
   ArrayList<ArrayList<Double>> resultA =  new ArrayList<>();


   public IncrementalSearch(){
       this.function = "";
       this.x0 = new BigDecimal(0.0);
       this.delta = 0.0;
       this.result = new double[]{0.0};
       this.niter = 0;
   }

   public IncrementalSearch(String function, BigDecimal x0, double delta, int niter){
    this.function = function;
    this.x0 = x0;
    this.delta = delta;
    this.result = new double[]{0.0};;
    this.niter = niter;
   }

   public String getMessage(){
       return this.message;
   }

  
    public ArrayList<ArrayList<Double>> incrementalSearch(){

        Expression expression = new Expression(this.function).setPrecision(10);
        expression.setVariable("x", this.x0);
        BigDecimal fx0 = expression.eval();
        BigDecimal x1;
        BigDecimal fx1;
        BigDecimal delta = new BigDecimal(this.delta);
        if (fx0.equals(0)) {

            ArrayList<Double> currentResult =  new ArrayList<>();
            currentResult.add(this.x0.doubleValue());
            this.resultA.add(currentResult);
                return this.resultA;
        } else {
            x1 = this.x0.add(delta);
            int i = 1;
            expression.setVariable("x", x1);
            fx1 = expression.eval();
            while ((fx0.doubleValue()*fx1.doubleValue() > 0) && (i < this.niter)) {
                ArrayList<Double> currentResult = new ArrayList();
                this.x0 = x1;
                fx0 = fx1;
                x1 = this.x0.add(delta);
                expression.setVariable("x", x1);
                fx1 = expression.eval();
                currentResult.add((double) i );
                currentResult.add(this.x0.doubleValue());
                currentResult.add(x1.doubleValue());
                currentResult.add(fx0.doubleValue());
                currentResult.add(fx1.doubleValue());
                this.resultA.add(currentResult);
                i++;                
            }
            if (fx1.doubleValue() == 0.0) {
                this.result = new double[]{x1.doubleValue()};
                this.message ="There is a root at " + x1.toString();
                return this.resultA;
            }
            else if (fx0.doubleValue()*fx1.doubleValue() < 0) {
                this.result = new double[]{this.x0.doubleValue(), x1.doubleValue()};
                this.message = "There is a root between " + this.x0.toString() + " and " +x1.toString();
                return this.resultA;
            } 
            else {
                System.out.println("Failed in " + i + "iterations");
                this.message = "Failed in " + i + " iterations";
                this.result = new double[]{-11111, -11111};
                return this.resultA;
            }   
        }
    }
}

File: M2M/app/src/main/java/FullPython/IncrementalSearch.py
from sympy import *
import math as m
from prettytable import PrettyTable

x = Symbol('x')

t = PrettyTable(['iteration','interval'])

def IncrementalSearch(tolerance,x0,delta,niter,function):
    fx0 = function.evalf(subs={x:x0})

    if m.fabs(fx0) < tolerance:
        print( str(x0)+" Is root")
    else:
        x1 = x0 + delta
        cont = 1
        fx1 = function.evalf(subs={x:x1})
        t.add_row([cont,'['+str(x1)+','+str(x0)+']'])
        while fx0 * fx1 > tolerance and cont< niter:
            x0 = x1
            fx0 = fx1
            x1 = x0 + delta
            fx1 = function.evalf(subs={x:x1})
            cont = cont + 1
            t.add_row([cont,'['+str(x0)+','+str(x1)+']'])
        print(t)
        if m.fabs(fx1) < tolerance:
            print (str(x1)+" Is root")
        elif fx0 * fx1 < tolerance:
            print ("There is a root between "+str(x0)+" and " + str(x1))
        else:
            print ("failed in  "+ str(niter)+ " iterations")

######################################
tolerance = 10**(-9)
x0 = -5
delta = 1
niter = 2000
function = (x**3) + (4*(x**(2)))-9.47
######################################

IncrementalSearch(tolerance,x0,delta,niter,function)
